[Executorch][target recipes] Add target based recipes for lowering models to a target device

abhinaykukkadapu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #13791

This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export
from executorch.export.target_recipes import get_ios_recipe

# CoreML + XNNPACK (FP32)
recipe = get_ios_recipe() # default = "ios-arm64-coreml-fp16"
session = export(model, recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage to combine specific recipes of one or two or backends
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(CoreMLRecipeType.FP32)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```
Additional changes:
1. Relaxed kwarg validation in the backend providers to just ignore them instead of erroring out.

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

abhinaykukkadapu committed:
[Executorch][target recipes] Add target based recipes for lowering models to a target device

This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export, ExportRecipe, IOSTargetRecipeType

# CoreML + XNNPACK
coreml_xnnpack_recipe = ExportRecipe.get_recipe(IOSTargetRecipeType.IOS_ARM64_COREML_FP32)
session = export(model, coreml_xnnpack_recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(AndroidRecipeType.XYZ)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export, ExportRecipe, IOSTargetRecipeType

# CoreML + XNNPACK
coreml_xnnpack_recipe = ExportRecipe.get_recipe(IOSTargetRecipeType.IOS_ARM64_COREML_FP32)
session = export(model, coreml_xnnpack_recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(AndroidRecipeType.XYZ)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export, ExportRecipe, IOSTargetRecipeType

# CoreML + XNNPACK
coreml_xnnpack_recipe = ExportRecipe.get_recipe(IOSTargetRecipeType.IOS_ARM64_COREML_FP32)
session = export(model, coreml_xnnpack_recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(AndroidRecipeType.XYZ)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export, ExportRecipe, IOSTargetRecipeType

# CoreML + XNNPACK
coreml_xnnpack_recipe = ExportRecipe.get_recipe(IOSTargetRecipeType.IOS_ARM64_COREML_FP32)
session = export(model, coreml_xnnpack_recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(AndroidRecipeType.XYZ)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu commented:
> I left a comment on the original issue. Not fully convinced why we need multibackend recipe as a separate concept

@kimishpatel  Removed the multi backend provider and went with simple funciton calls.

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export, ExportRecipe, IOSTargetRecipeType

# CoreML + XNNPACK
coreml_xnnpack_recipe = ExportRecipe.get_recipe(IOSTargetRecipeType.IOS_ARM64_COREML_FP32)
session = export(model, coreml_xnnpack_recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(AndroidRecipeType.XYZ)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

kimishpatel commented on backends/apple/coreml/recipes/coreml_recipe_provider.py:
```diff
@@ -111,8 +112,9 @@ def _validate_recipe_kwargs(self, recipe_type: RecipeType, **kwargs: Any) -> Non
         if unexpected:
-            raise ValueError(
-                f"Recipe '{recipe_type.value}' received unexpected parameters: {list(unexpected)}"
+            logging.warning(
```
Why this turned into warning now? to combine multiple recipes? and fallback to the other backend when it fails?

kimishpatel commented on export/recipe.py:
```diff
@@ -184,3 +185,126 @@ def get_recipe(cls, recipe: "RecipeType", **kwargs) -> "ExportRecipe":
+
+        Args:
+            recipes: List of ExportRecipe objects to combine
+            recipe_name: Optional name for the combined recipe
```
why not just combine it from, the name of the underlying recipes

kimishpatel reviewed (commented):
looks ok to me. some nits plus a comment on quantizaiton.

you probably want to add recipes plus test for android

abhinaykukkadapu commented on export/recipe.py:
```diff
@@ -184,3 +185,126 @@ def get_recipe(cls, recipe: "RecipeType", **kwargs) -> "ExportRecipe":
+
+            # Collect for quantize stage
+            if quantization_recipe := recipe.quantization_recipe:
+                # Collect PT2E quantizers
+                if quantization_recipe.quantizers:
+                    all_quantizers.extend(quantization_recipe.quantizers)
```
This is just to compose it as a list for ExportRecipe object, in the actual execution, we do use composable quantizer: [main/export/stages.py#L365](https://github.com/pytorch/executorch/blob/main/export/stages.py#L365)

abhinaykukkadapu replied:
Gotcha, thanks for pointing it, will disable in next version.

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export, ExportRecipe, IOSTargetRecipeType

# CoreML + XNNPACK
coreml_xnnpack_recipe = ExportRecipe.get_recipe(IOSTargetRecipeType.IOS_ARM64_COREML_FP32)
session = export(model, coreml_xnnpack_recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(AndroidRecipeType.XYZ)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export
from executorch.export.target_recipes import get_ios_recipe

# CoreML + XNNPACK (FP32)
recipe = get_ios_recipe() # default = "ios-arm64-coreml-fp16"
session = export(model, recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage to combine specific recipes of one or two or backends
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(CoreMLRecipeType.FP32)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```
Additional changes:
1. Relaxed kwarg validation in the backend providers to just ignore them instead of erroring out.

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export
from executorch.export.target_recipes import get_ios_recipe

# CoreML + XNNPACK (FP32)
recipe = get_ios_recipe() # default = "ios-arm64-coreml-fp16"
session = export(model, recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage to combine specific recipes of one or two or backends
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(CoreMLRecipeType.FP32)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```
Additional changes:
1. Relaxed kwarg validation in the backend providers to just ignore them instead of erroring out.

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export
from executorch.export.target_recipes import get_ios_recipe

# CoreML + XNNPACK (FP32)
recipe = get_ios_recipe() # default = "ios-arm64-coreml-fp16"
session = export(model, recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage to combine specific recipes of one or two or backends
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(CoreMLRecipeType.FP32)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```
Additional changes:
1. Relaxed kwarg validation in the backend providers to just ignore them instead of erroring out.

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[Executorch][target recipes] Add target based recipes for lowering models to a target device"


This diff introduces multi backend/ target based recipes to lower a model with very less code. Target recipes provide pre-configured backend recipes to use them and retarget if needed.

See RFC: https://github.com/pytorch/executorch/issues/13732

## Usage

```
from executorch.export import export
from executorch.export.target_recipes import get_ios_recipe

# CoreML + XNNPACK (FP32)
recipe = get_ios_recipe() # default = "ios-arm64-coreml-fp16"
session = export(model, recipe, example_inputs)
session.save_pte_file("model.pte")
```

## Advanced usage to combine specific recipes of one or two or backends
one can directly use `ExportRecipe.combine_recipes()` to combine specific backend recipes.
```
recipe1 = ExportRecipe.get_recipe(CoreMLRecipeType.FP32)
recipe2 = ExportRecipe.get_recipe(XNNPackRecipeType.FP32)
combined_recipe = ExportRecipe.combine(
                     [recipe1, recipe2],
                     recipe_name="multi_backend_coreml_xnnpack_fp32"
                  )
session = export(model, combined_recipe, example_inputs)
```
Additional changes:
1. Relaxed kwarg validation in the backend providers to just ignore them instead of erroring out.

Fixes: #13732

Differential Revision: [D81297451](https://our.internmc.facebook.com/intern/diff/D81297451/)

[ghstack-poisoned]
